### PR TITLE
[WIP] Computes training loss for skip gram model. Fixes issue #999.

### PR DESCRIPTION
**Pull request by chinmayapancholi13**
This PR computes the training loss for each pair of words being trained as per the skip gram model. The loss value is computed in the function `train_sg_pair` and the value is displayed after every `print_freq` (passed as a parameter to function `train_batch_sg`) number of training pairs. To not show the loss value at all, set `print_freq` as 0 (which is the default value).

### CODE REVIEW

**Comment by piskvorky** [comment]
I can't see how a pure-Python implementation would be useful... I assume people who need to see progress need to see it on large data. And pure-Python on large data would be too slow to start with.

**Comment by chinmayapancholi13** [comment]
@piskvorky Yes I agree completely. However, the manner in which we want to compute and use the loss value is itself not clear right now (ongoing discussion at https://github.com/RaRe-Technologies/gensim/issues/999). This is why I had submitted a PR first by making changes for the Python path for skip gram model and have planned to make appropriate changes for other paths, both for skip gram and CBOW, for Word2Vec (and also for Doc2Vec).

File: gensim/models/word2vec.py [hunk 1]
```diff
@@ -140,7 +140,7 @@
     FAST_VERSION = -1
     MAX_WORDS_IN_BATCH = 10000
 
-    def train_batch_sg(model, sentences, alpha, work=None):
+    def train_batch_sg(model, sentences, alpha, work=None, print_freq=0):
```
Comment (menshikh-iv) [inline gensim/models/word2vec.py]: Maybe rename `print_freq`  to `eval_every` as well as in `LdaModel`

**Comment by menshikh-iv** [comment]
A good start, I think. Although pure python implementation can be slow (need the benchmark for check this), if you rarely logging model loss, this does not slow learning process significantly (but Cython version is the best choice).

File: gensim/models/word2vec.py (hunk 1 above)
Comment (chinmayapancholi13) [inline gensim/models/word2vec.py]: Sure Ivan. I'll update the parameter name.

**Comment by gojomo** [comment]
I personally don't see this current implementation as useful enough to justify the code-complication. The people who requested it likely don't use the pure-python path – it's an advanced need, and advanced users are dependent on the optimized code.

The discussion on #999 seemed to conclude that accumulating the error in the model was a superior approach – but this only logs the value of `train_error_value`, the error of a single (!) skip-pair training example, occasionally. (I think that number will jump all over, based on which 2 words happen to have been in the example when logging is chosen. I find it hard to imagine any use for such logged values.)

I believe this feature will require the active participation of someone who actively needs the feature – either to provide a good spec & review, or to implement – to become a worthwhile addition.

**Comment by chinmayapancholi13** [comment]
@gojomo I am indeed planning to implement this for both - python path as well as the optimized cython path.
Also, as per the latest discussion on https://github.com/RaRe-Technologies/gensim/issues/999, it was agreed that storing the *sum of the training losses of all the word-pairs* would be a good start to address the issue. As mentioned in the comment by @dietmar [here](https://github.com/RaRe-Technologies/gensim/issues/999#issuecomment-288983230), this would enable a user to get the *cumulative* training loss value for each call to the function `train()`.

So overall this should infact assist the users in getting an idea about how the training of their model is progressing. But having said that, I do agree that this may actually be just one of the many use-cases that users might need this functionality for so more user input before going forward with the implementation wouldn't hurt.

cc : @tmylk @menshikh-iv

**Comment by gojomo** [comment]
At the time of my comment, before commit 8949749, there wasn't yet any accumulation – only logging of occasional single-pair errors. Yes, a value for a whole epoch, that is thus comparable against another epoch, would be useful. (A running average of 'recent' error might also be useful.)

**Comment by piskvorky** [comment]
How does this extra branching / extra code (caching effects) affect performance?

@chinmayapancholi13 Can you post some benchmarks, before + after?

**Comment by chinmayapancholi13** [comment]
@piskvorky Sure. I plan to post benchmarks for before v/s after cases, as soon as I complete the coding part.

File: gensim/test/test_word2vec.py [hunk 2]
```diff
@@ -698,6 +698,13 @@ def test_reset_from(self):
         model.reset_from(other_model)
         self.assertEqual(model.wv.vocab, other_vocab)
 
+    def test_compute_training_loss(self):
+        model = word2vec.Word2Vec(min_count=1, sg=1, negative=5, hs=1)
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: Please add tests for more training modes.

**Comment by chinmayapancholi13** [comment]
**Benchmarks for seeing effect of loss computation code on training time**
text8_50000000 : 50 MB size
lee_background.cor : 25 KB size

train_data | compute_loss | hs | sg | mean | std | 
| --- | --- | --- | --- | --- | --- |
text8_50000000 | True | 0 | 1 | 125.242767 | 0.442522 |    
text8_50000000 | False | 0 | 1 | 124.090732 | 1 | 0.432895 |    
text8_50000000 | True | 1 | 1 | 252.800164 | 1.140344 |    
text8_50000000 | False | 1 | 1 | 245.065643 | 2.657392 |    
text8_50000000 | True | 0 | 0 | 43.812430 | 1.216697 |    
text8_50000000 | False | 0 | 0 | 42.815214 | 0.142814 |    
text8_50000000 | True | 1 | 0 | 74.801153 | 0.300728 |    
text8_50000000 | False | 1 | 0 | 74.236441 | 0.126426 |    
lee_background.cor | True | 0 | 1 | 0.560387 | 0.005805 |    
lee_background.cor | False | 0 | 1 | 0.687179 | 0.143629 |    
lee_background.cor | True | 1 | 1 | 1.126855 | 0.004407 |    
lee_background.cor | False | 1 | 1 | 1.135358 | 0.059161 |    
lee_background.cor | True | 0 | 0 | 0.316948 | 0.005148 |    
lee_background.cor | False | 0 | 0 | 0.319236 | 0.005792 |    
lee_background.cor | True | 1 | 0 | 0.429879 | 0.005373 |    
lee_background.cor | False | 1 | 0 | 0.429489 | 0.000756 |

**Comment by gojomo** [comment]
The lee corpus is too small/quick to get any feedback. From the text8 results, it looks like 'true' may cause an 0.5% to 3% slowdown. A comparison against the code without even the switch-to-turn-it-on (without this feature at all) might be relevant.

File: continuous_integration/travis/flake8_diff.sh [hunk 3]
```diff
@@ -133,6 +133,6 @@ check_files() {
 if [[ "$MODIFIED_FILES" == "no_match" ]]; then
     echo "No file has been modified"
 else
-    check_files "$(echo "$MODIFIED_FILES" )" "--ignore=E501,E731,E12,W503 --exclude=*.sh,*.md,*.yml,*.rst,*.ipynb,*.txt,*.csv,*.vec,Dockerfile*"
+    check_files "$(echo "$MODIFIED_FILES" )" "--ignore=E501,E731,E12,W503 --exclude=*.sh,*.md,*.yml,*.rst,*.ipynb,*.txt,*.csv,*.vec,Dockerfile*,*.c,*.pyx"
```
Comment (gojomo) [inline continuous_integration/travis/flake8_diff.sh]: Are we sure `.pyx` should be here? I didn't see what kind of warnings flake was generating, but as cython syntax is mostly python, and most of our enforceable conventions should still be in effect, we may want some style-enforcement there.

**Comment by gojomo** [comment]
While summing all the errors for a single call to `train()` is somewhat more useful, to a knowledgeable user, than the prior implements, it's still not quite what users may find interpretable. (At least not compared to the running-average error printed during training by other tools like fasttext.) I also think it might need to be `abs()`d somewhere, lest offsetting errors in both directions be misleading when totaled. So I don't really see this as ready-for-merging, again preferring someone who actively needs the feature to review it.

File: continuous_integration/travis/flake8_diff.sh (hunk 3 above)
Comment (menshikh-iv) [inline continuous_integration/travis/flake8_diff.sh]: @gojomo flake8 can't correctly check `pyx` files

Comment (chinmayapancholi13) [inline continuous_integration/travis/flake8_diff.sh]: @gojomo We were getting errors like these due to flake8 : 
flake8_2

flake8_1

So although I do agree that there is some style-checking that we might want to do in .pyx files (in the python-like code), to avoid getting errors due to cases similar to the above cases, I thought it would be better to ignore .pyx for flake8 tests.

**Comment by piskvorky** [comment]
What does `mean` and `std` mean in the table? Mean and std of what?

I'm most interested in seeing a performance comparison against the original. Comparison of the new switch on/off is interesting too, but we have to evaluate what we're losing by adding an extra switch to the critical path.

And like @gojomo says, it's best to run the benchmark against a larger corpus and multiple times, for robustness.

**Comment by chinmayapancholi13** [comment]
@piskvorky `mean` and `std` refer to the average value and standard deviation of the time taken for training the word2vec model.

When we say a "larger corpus", could you please give an approx size? As I have mentioned in my comment [here](https://github.com/RaRe-Technologies/gensim/pull/1201#issuecomment-312038344), I did this comparison between the old and new training times for 100 MB text8 corpus as well (but not for all the possible 8 combinations of `compute_loss`, `sg` and `hs` values) and I was getting similar results (i.e. the difference in time values was about the same percentage as the results posted for 50 MB test data).

File: continuous_integration/travis/flake8_diff.sh (hunk 3 above)
Comment (gojomo) [inline continuous_integration/travis/flake8_diff.sh]: Ah, I see. There's an SO answer that implies it may be possible to turn off just certain warnings for `.pyx` files – https://stackoverflow.com/questions/31269527/running-pep8-or-pylint-on-cython-code – though the full example file is a broken link.

**Comment by gojomo** [comment]
I'd find a couple tests that runs for tens-of-minutes better than many tests that run for a couple-minutes (or just tens-of-seconds). Increasing the `iter` to 20 or 40 might achieve that with text8. Also, if setting `hs=1` should also set `negative=0` (to match usual/recommended user choices of jsut one mode or the other – even though it *is* possible to run both at the same time and thus be benchmarking the combination).

File: continuous_integration/travis/flake8_diff.sh (hunk 3 above)
Comment (chinmayapancholi13) [inline continuous_integration/travis/flake8_diff.sh]: Thanks for sharing this link. :) I can try to use the config specified in the answer here to check if all the undesired warnings/errors are turned off using it.

**Comment by chinmayapancholi13** [comment]
Got it, @gojomo. So to conclude, we should use the text8 corpus with `iter` set around 20-40 and compare the time taken by the new code with `compute_loss` set to `True` with the original code (without any switch). I'll update the benchmarks as soon as I have these values. :)

**Comment by chinmayapancholi13** [comment]
@gojomo @piskvorky 

**Analysis of the extra time taken for training due to loss compuation code** 

**Note 1** :  Here, the rows corresponding to `compute_loss=False` use the old code (without any `compute_loss` switch)
**Note 2** : Machine specs : 64 GB RAM, i-7 processor, 3.38 GHz (upto 3.6 GHz) 

|compute_loss|     hs|	   sg |    mean |       std |
| --- | --- | --- | --- | --- |
|True  		|	1 	|	1 | 341.065673 |  0.722949   
|False  		|	1 	|	1 | 338.080630 | 0.541960   
|True  		|	0 	|	1 | 172.808942 | 2.547450   
|False  		|	0	|	1 | 169.936202 | 0.422062   
|True  		|	1 	|	0 | 103.503094  | 0.930082   
|False  		|	1 	|	0 | 102.959268  | 1.517684
|True  		|	0 	|	0 | 58.828162   | 0.410273   
|False  		|	0 	|	0 | 59.028026   | 0.161999

**Comment by gojomo** [comment]
The most interesting comparison, to determine the overhead for people who won't us the feature, would be between the old code (no changes present), and the new code with `compute_loss=False`.

**Comment by chinmayapancholi13** [comment]
@gojomo I agree. I have updated the table to include these values too.

**Comment by gojomo** [comment]
AFAICT, whatever overhead there is from both the switch, or enabling the switch, disappears into the variance (in these short tests), so there's no evidence of a performance-related reason to hold back. 

But for the reasons in my comment https://github.com/RaRe-Technologies/gensim/pull/1201#issuecomment-312054254 I still think this was merged (and #999 closed) prematurely, because there's no confirmation of usefulness from the people who have actually requested this. 

Also, while the titles still mention "skip-gram", the benchmarking included non-skip-gram-modes... so does that mean what's been done works for CBOW, too?

**Comment by chinmayapancholi13** [comment]
@gojomo Since the title https://github.com/RaRe-Technologies/gensim/issues/999 mentions "skip-gram", this PR was originally only for skip-gram mode but later I decided to implement both modes in the same PR. Thanks for pointing this out. :) I have updated the title of the PR now.
